File: backend/app/Modules/Riding/StateMachine/RideStatus.php
<?php

declare(strict_types=1);

namespace App\Modules\Riding\StateMachine;

enum RideStatus: string
{
    case Requested = 'requested';
    case Searching = 'searching';
    case Offered = 'offered';
    case Accepted = 'accepted';
    case DriverArriving = 'driver_arriving';
    case DriverArrived = 'driver_arrived';
    case InProgress = 'in_progress';
    case Completed = 'completed';
    case Cancelled = 'cancelled';
    case NoDrivers = 'no_drivers';
    case Failed = 'failed';

    public function isTerminal(): bool
    {
        return in_array($this, [self::Completed, self::Cancelled, self::NoDrivers, self::Failed], true);
    }

    public function isActive(): bool
    {
        return ! $this->isTerminal();
    }
}

File: mobile/packages/rides/lib/src/models/ride_status.dart
/// Mirror of `App\Modules\Riding\StateMachine\RideStatus`. Keep in sync
/// with the backend enum.
enum RideStatus {
  requested,
  searching,
  offered,
  accepted,
  driverArriving,
  driverArrived,
  inProgress,
  completed,
  cancelled,
  noDrivers,
  failed;

  static RideStatus fromWire(String value) => switch (value) {
        'requested' => RideStatus.requested,
        'searching' => RideStatus.searching,
        'offered' => RideStatus.offered,
        'accepted' => RideStatus.accepted,
        'driver_arriving' => RideStatus.driverArriving,
        'driver_arrived' => RideStatus.driverArrived,
        'in_progress' => RideStatus.inProgress,
        'completed' => RideStatus.completed,
        'cancelled' => RideStatus.cancelled,
        'no_drivers' => RideStatus.noDrivers,
        'failed' => RideStatus.failed,
        _ => RideStatus.failed,
      };

  bool get isTerminal => switch (this) {
        RideStatus.completed ||
        RideStatus.cancelled ||
        RideStatus.noDrivers ||
        RideStatus.failed =>
          true,
        _ => false,
      };

  bool get isPreAssignment => switch (this) {
        RideStatus.requested || RideStatus.searching || RideStatus.offered => true,
        _ => false,
      };

  bool get hasDriver => switch (this) {
        RideStatus.accepted ||
        RideStatus.driverArriving ||
        RideStatus.driverArrived ||
        RideStatus.inProgress =>
          true,
        _ => false,
      };
}
